source/dubregistry/registry.d

```diff
@@ -575,4 +628,48 @@ struct PackageVersionInfo {
 struct PackageInfo {
 	PackageVersionInfo[] versions;
 	Json info; /// JSON package information, as reported to the client
-}
\ No newline at end of file
+}
+
+/// Computes a package rating from given package stats and global distributions of those stats.
+private float computeRating(DownDist, RepoDist)(in ref DbPackageStats stats, DownDist downDist, RepoDist repoDist)
+{
+	import std.math : log1p, round, tanh;
+
+    if (!downDist.total.sum) // no stat distribution yet
+        return 0;
+
+	/// Using monthly downloads to penalize stale packages, logarithm to
+	/// offset exponential distribution, and tanh as smooth limiter to [0..1].
+	immutable downloadRating = tanh(log1p(stats.downloads.monthly / downDist.monthly.mean));
+	logDebug("downloadRating %s %s %s", downloadRating, stats.downloads.monthly, downDist.monthly.mean);
+
+	// Compute rating for repo
+	float sum=0, wsum=0;
+	void add(T)(float weight, float value, T dist)
+	{
+		if (dist.sum == 0)
+			return; // ignore metrics missing for that repository kind
+		sum += weight * log1p(value / dist.mean);
+		wsum += weight;
+	}
+	with (stats.repo)
+	{
+		alias d = repoDist;
+		// all of those values are highly correlated
+		add(1.0f, stars, d.stars);
+		add(1.0f, watchers, d.watchers);
+		add(1.0f, forks, d.forks);
+		add(-1.0f, issues, d.issues); // penalize many open issues/PRs
+	}
+
+	immutable repoRating = max(0.0, tanh(sum / wsum));
+	logDebug("repoRating: %s %s %s", repoRating, sum, wsum);
+
+	// average ratings
+	immutable avgRating = (repoRating + downloadRating) / 2;
+	assert(0 <= avgRating && avgRating <= 1.0, "%s %s".format(repoRating, downloadRating));
+	immutable scaled = stats.minRating + avgRating * (stats.maxRating - stats.minRating);
+	logDebug("rating: %s %s %s %s %s %s", stats.downloads.monthly, downDist.monthly.mean, downloadRating, repoRating, avgRating, scaled);
+
+	return scaled;
```

Comment:
Cool, would be great to add a comment that summarizes the computation in somewhat math/natural language notation, e.g. "weighted sum of stars, watches, forks with all weights 1.0, minus penalty for issues weighted by 1.0 each" etc.

I assume there will be some tweaking over time.